[CUSPARSE] Add CuSparseVector ± CuSparseVector

amontoison opened the pull request:
@maleadt 
It should solve the issue on Slack.

amontoison committed:
[CUSPARSE] Add CuSparseVector ± CuSparseVector

maleadt commented on lib/cusparse/extra.jl:
```diff
@@ -19,41 +26,81 @@ for (bname,gname,elty) in ((:cusparseScsrgeam2_bufferSizeExt, :cusparseScsrgeam2
                     alpha, descrA, nnz(A), nonzeros(A), A.rowPtr, A.colVal,
                     beta, descrB, nnz(B), nonzeros(B), B.rowPtr, B.colVal,
-                    descrC, CuArray{$elty,1}(undef, 0), rowPtrC, CuArray{Int32,1}(undef, 0),
+                    descrC, CuPtr{$elty}(CU_NULL), rowPtrC, CuPtr{Int32}(CU_NULL),
```
Same comment here as on the other PR.

amontoison committed:
Don't convert CU_NULL

maleadt commented:
Nice! I'm not entirely sure about implementing functions like `geam` for types that CUSPARSE doesn't actually support, as they're still intended as fairly low-level wrappers. Something like reinterpreting the CuSparseVector as a 1-row CuSparseMatrixCSR and re-using the existing `geam` methods seems cleaner in that regard. But I'll let you decide what's best, since you've been spending more time working on the CUSPARSE wrappers than I ever have 🙂

amontoison committed:
Update the function for CuSparseVector ± CuSparseVector

amontoison commented on lib/cusparse/extra.jl:
```diff
@@ -14,7 +14,7 @@ for (bname,gname,elty) in ((:cusparseScsrgeam2_bufferSizeExt, :cusparseScsrgeam2
         function geam(alpha::Number, A::CuSparseMatrixCSR{$elty}, beta::Number, B::CuSparseMatrixCSR{$elty}, index::SparseChar)
             m, n = size(A)
-            (m, n) == size(B) && DimensionMismatch("dimensions must match: a has dims $(axes(A)), b has dims $(axes(B))")
+            (m, n) == size(B) || DimensionMismatch("dimensions must match: A has dims $(size(A)), B has dims $(size(B))")
```
@maleadt 
Do you have an idea why it was working before. :thinking:

amontoison replied:
Update: a `throw` was missing...

maleadt replied:
Haha, nice catch!

amontoison committed:
Update the function for CuSparseVector ± CuSparseVector

amontoison commented:
> Nice! I'm not entirely sure about implementing functions like `geam` for types that CUSPARSE doesn't actually support, as they're still intended as fairly low-level wrappers. Something like reinterpreting the CuSparseVector as a 1-row CuSparseMatrixCSR and re-using the existing `geam` methods seems cleaner in that regard. But I'll let you decide what's best, since you've been spending more time working on the CUSPARSE wrappers than I ever have slightly_smiling_face

You are right, we should reinterprete the CuSparseVector as a 1-row CuSparseMatrixCSR and reuse the existing `geam`.
I updated the code and also renamed the `geam` like function for CuSparseVector `axpby`.

amontoison committed:
Add a missing throw

maleadt commented:
I like that better, thanks!
